pkg/identity/ciprovider/principal.go

```diff
@@ -0,0 +1,141 @@
+// Copyright 2024 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ciprovider
+
+import (
+	"bytes"
+	"context"
+	"crypto/x509"
+	"fmt"
+	"html/template"
+	"net/url"
+	"strings"
+
+	"github.com/coreos/go-oidc/v3/oidc"
+	"github.com/sigstore/fulcio/pkg/certificate"
+	"github.com/sigstore/fulcio/pkg/config"
+	"github.com/sigstore/fulcio/pkg/identity"
+)
+
+func claimsToString(claims map[string]interface{}) map[string]string {
+	stringClaims := make(map[string]string)
+	for k, v := range claims {
+		stringClaims[k] = v.(string)
+	}
+	return stringClaims
+}
+
+// It makes string interpolation for a given string by using the
+// templates syntax https://pkg.go.dev/text/template
+func applyTemplateOrReplace(path string, data map[string]string, defaultData map[string]string) string {
+
+	// Here we merge the data from was claimed by the id token with the
+	// default data provided by the yaml file.
+	// The order here matter because we want to override the default data
```

Comment:
I think we want the reverse order here. The values provided as defaults are supplemental, as in we don't expect `url` to be set as a value in the token. I don't think we should err out if `url` was set in both because we don't control the contents of the token, but we should specify in the configuration set up to choose additional values where those keys won't be present in the token.

If they are present, then we should choose what's in `defaultData` over `data`.

WDYT?

Reply:
It makes sense!